server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java

```diff
@@ -196,6 +196,29 @@ ClusterState addComponentTemplate(final ClusterState currentState, final boolean
 
         validateTemplate(finalSettings, stringMappings, indicesService, xContentRegistry);
 
+        // if we're updating a component template, let's check if it's part of any V2 template that will yield the CT update invalid
+        if (create == false && finalSettings != null) {
+            // if the CT is specifying the `index.hidden` setting it cannot be part of any global template
+            if (IndexMetadata.INDEX_HIDDEN_SETTING.exists(finalSettings)) {
+                Map<String, IndexTemplateV2> existingTemplates = currentState.metadata().templatesV2();
+                List<String> globalTemplatesThatUseThisComponent = new ArrayList<>();
+                for (Map.Entry<String, IndexTemplateV2> entry : existingTemplates.entrySet()) {
+                    IndexTemplateV2 templateV2 = entry.getValue();
+                    if (templateV2.composedOf().contains(name) && templateV2.indexPatterns().stream().anyMatch(Regex::isMatchAllPattern)) {
+                        // global templates don't support configuring the `index.hidden` setting so we don't need to resolve the settings as
+                        // no other component template can remove this setting from the resolved settings, so just invalidate this update
+                        globalTemplatesThatUseThisComponent.add(entry.getKey());
+                    }
+                }
```

Comment:
This can be replaced with stream version, which for me is more readable:
```java
List<String> globalTemplatesThatUseThisComponent = currentState.metadata().templatesV2().entrySet().stream()
                .filter(e -> e.getValue().composedOf().contains(name))
                .filter(e -> e.getValue().indexPatterns().stream().anyMatch(Regex::isMatchAllPattern))
                .map(Map.Entry::getKey)
                .collect(Collectors.toList());
```

Reply:
I'm not sure about this, to me they're equally readable, but your option is a bit more concise. 
In the interest of moving this rather stale PR forward I'm merging it as it is and we can refactor it later if that's ok?